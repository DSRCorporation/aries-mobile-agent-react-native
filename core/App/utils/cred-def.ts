import { CredentialMetadataKeys, CredentialExchangeRecord as CredentialRecord } from '@aries-framework/core'

import { parseSchema, parseSchemaFromId, credentialSchema } from './schema'

export function parseCredDefFromId(credDefId?: string, schemaId?: string): string {
  let name = 'Credential'
  if (credDefId) {
    const credDefRegex = /[^:]+/g
    const credDefIdParts = credDefId.match(credDefRegex)
    if (credDefIdParts?.length === 5) {
      name = `${credDefIdParts?.[4].replace(/_|-/g, ' ')}`
        .split(' ')
        .map((credIdPart) => credIdPart.charAt(0).toUpperCase() + credIdPart.substring(1))
        .join(' ')
    }
  }
  if (name.toLocaleLowerCase() === 'default' || name.toLowerCase() === 'credential') {
    name = parseSchemaFromId(schemaId).name
  }
  return name
}

function credentialDefinition(credential: CredentialRecord): string | undefined {
  return credential.metadata.get(CredentialMetadataKeys.IndyCredential)?.credentialDefinitionId
}

export function parsedCredDefNameFromCredential(credential: CredentialRecord): string {
  return parseCredDefFromId(credentialDefinition(credential), credentialSchema(credential))
}

export function parsedCredDefName(credentialDefinitionId: string, schemaId: string): string {
  let credDefName = parseCredDefName(credentialDefinitionId)
  // if credDef name is `default` or `credential` use schema name instead
  if (credDefName.toLocaleLowerCase() === 'default' || credDefName.toLowerCase() === 'credential') {
    credDefName = parseSchema(schemaId).name
  }
  return credDefName
}
