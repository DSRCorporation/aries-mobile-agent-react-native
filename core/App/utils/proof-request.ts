import { Agent, AgentMessage, AutoAcceptProof, ProofExchangeRecord } from '@aries-framework/core'

import { defaultProofRequestTemplates } from '../constants'
import {
  IndyRequestedAttribute,
  IndyRequestedPredicate,
  ProofRequestTemplate,
  ProofRequestType,
} from '../types/proof-reqeust-template'

const protocolVersion = 'v1'
const domain = 'http://aries-mobile-agent.com'

/*
 * Find Proof Request template for provided id
 * */
export const getProofRequestTemplate = (id: string): ProofRequestTemplate | undefined => {
  return defaultProofRequestTemplates.find((template) => template.id === id)
}

/*
 * Build Proof Request data from for provided template
 * */
export const buildProofRequestDataForTemplate = (template: ProofRequestTemplate) => {
  if (template.payload.type === ProofRequestType.Indy) {
    const requestedAttributes: Map<string, IndyRequestedAttribute> = new Map()
    const requestedPredicates: Map<string, IndyRequestedPredicate> = new Map()
    let index = 0

    template.payload.data.forEach((data) => {
      if (data.requestedAttributes?.length) {
        data.requestedAttributes.forEach((requestedAttribute) => {
          requestedAttributes.set(`referent_${index}`, requestedAttribute)
          index++
        })
      }
      if (data.requestedPredicates?.length) {
        data.requestedPredicates.forEach((requestedPredicate) => {
          requestedPredicates.set(`referent_${index}`, requestedPredicate)
          index++
        })
      }
    })
    return {
      indy: {
        name: template.title,
        version: template.version,
        nonce: Date.now().toString(),
        requestedAttributes,
        requestedPredicates,
      },
    }
  }
  if (template.payload.type === ProofRequestType.DIF) {
    return {}
  }
}

export const buildProofRequestDataForTemplateId = (templateId: string) => {
  const template = getProofRequestTemplate(templateId)
  if (!template) {
    return undefined
  }
  return buildProofRequestDataForTemplate(template)
}

export interface CreateProofRequestInvitationResult {
  request: AgentMessage
  proofRecord: ProofExchangeRecord
  invitation: AgentMessage
  invitationUrl: string
}

/*
 * Create connectionless proof request invitation for provided template
 * */
export const createConnectionlessProofRequestInvitation = async (
  agent: Agent,
  templateId: string
): Promise<CreateProofRequestInvitationResult | undefined> => {
  const proofFormats = buildProofRequestDataForTemplateId(templateId)
  if (!proofFormats) {
    return undefined
  }
  const { message: request, proofRecord } = await agent.proofs.createRequest({
    protocolVersion,
    autoAcceptProof: AutoAcceptProof.Always,
    proofFormats,
  })
  const { message: invitation, invitationUrl } = await agent.oob.createLegacyConnectionlessInvitation({
    recordId: proofRecord.id,
    message: request,
    domain,
  })
  return {
    request,
    proofRecord,
    invitation,
    invitationUrl,
  }
}

export interface SendProofRequestResult {
  proofRecord: ProofExchangeRecord
}

/*
 * Build Proof Request for provided template and send it to provided connection
 * */
export const sendProofRequest = async (
  agent: Agent,
  templateId: string,
  connectionId: string
): Promise<SendProofRequestResult | undefined> => {
  const proofFormats = buildProofRequestDataForTemplateId(templateId)
  if (!proofFormats) {
    return undefined
  }
  const proofRecord = await agent.proofs.requestProof({
    protocolVersion,
    connectionId,
    proofFormats,
  })
  return {
    proofRecord,
  }
}
