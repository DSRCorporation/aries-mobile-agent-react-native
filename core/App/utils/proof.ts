import { IndyProof, IndyProofRequest } from 'indy-sdk-react-native'

export interface MissingAttribute {
  name: string
}

export interface SharedAttribute {
  name: string
  value: string
}

export interface ResolvedPredicate {
  name: string
  predicateType: string
  predicateValue: number
}

export interface ParsedIndyProof {
  sharedAttribute: Array<SharedAttribute>
  sharedAttributeGroups: Array<Array<SharedAttribute>>
  resolvedPredicates: Array<ResolvedPredicate>
  unresolvedAttributes: Array<MissingAttribute>
  unresolvedAttributeGroups: Array<Array<MissingAttribute>>
  unresolvedPredicates: Array<ResolvedPredicate>
}

export const parseIndyProof = (request: IndyProofRequest, proof: IndyProof) => {
  const sharedAttribute: Array<SharedAttribute> = []
  const sharedAttributeGroups: Array<Array<SharedAttribute>> = []
  const resolvedPredicates: Array<ResolvedPredicate> = []
  const unresolvedAttributes: Array<MissingAttribute> = []
  const unresolvedAttributeGroups: Array<Array<MissingAttribute>> = []
  const unresolvedPredicates: Array<ResolvedPredicate> = []

  for (const [referent, requested_attribute] of Object.entries(request.requested_attributes)) {
    if (requested_attribute.name) {
      const shared = proof.requested_proof.revealed_attrs[referent]
      if (shared) {
        sharedAttribute.push({
          name: requested_attribute.name,
          value: shared.raw,
        })
      } else {
        unresolvedAttributes.push({
          name: requested_attribute.name,
        })
      }
    }

    if (requested_attribute.names) {
      const shared = proof.requested_proof.revealed_attr_groups[referent]
      if (shared) {
        sharedAttributeGroups.push(
          Object.entries(shared.values).map(([name, value]) => ({
            name,
            value: value.raw,
          }))
        )
      } else {
        unresolvedAttributeGroups.push(requested_attribute.names.map((name) => ({ name })))
      }
    }
  }

  for (const [referent, requestedPredicate] of Object.entries(request.requested_predicates)) {
    const shared = proof.requested_proof.requested_predicates[referent]
    if (shared) {
      resolvedPredicates.push({
        name: requestedPredicate.name,
        predicateType: requestedPredicate.p_type,
        predicateValue: requestedPredicate.p_value,
      })
    } else {
      unresolvedPredicates.push({
        name: requestedPredicate.name,
        predicateType: requestedPredicate.p_type,
        predicateValue: requestedPredicate.p_value,
      })
    }
  }

  return {
    sharedAttribute,
    sharedAttributeGroups,
    resolvedPredicates,
    unresolvedAttributes,
    unresolvedAttributeGroups,
    unresolvedPredicates,
  }
}
