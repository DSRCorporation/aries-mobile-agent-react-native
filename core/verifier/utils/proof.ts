import { Agent, ProofExchangeRecord, ProofIdentifier, ProofState } from '@aries-framework/core'
import { IndyProof, IndyProofRequest } from 'indy-sdk-react-native'

export interface MissingAttribute {
  name: string
}

export interface SharedAttribute {
  name: string
  value: string
  identifiers: ProofIdentifier
}

export interface SharedGroupedAttribute {
  name: string
  value: string
}

export interface SharedAttributesGroup {
  attributes: Array<SharedGroupedAttribute>
  identifiers: ProofIdentifier
}

export interface ResolvedPredicate {
  name: string
  predicateType: string
  predicateValue: number
  identifiers: ProofIdentifier
}

export interface UnresolvedPredicate {
  name: string
  predicateType: string
  predicateValue: number
}

export class ParsedIndyProof {
  public sharedAttributes: Array<SharedAttribute>
  public sharedAttributeGroups: Array<SharedAttributesGroup>
  public resolvedPredicates: Array<ResolvedPredicate>
  public unresolvedAttributes: Array<MissingAttribute>
  public unresolvedAttributeGroups: Array<Array<MissingAttribute>>
  public unresolvedPredicates: Array<UnresolvedPredicate>

  public constructor() {
    this.sharedAttributes = []
    this.sharedAttributeGroups = []
    this.resolvedPredicates = []
    this.unresolvedAttributes = []
    this.unresolvedAttributeGroups = []
    this.unresolvedPredicates = []
  }
}

/*
 * Extract identifiers from indy proof
 * */
const getProofIdentifiers = (proof: IndyProof, proofIndex: number): ProofIdentifier => {
  const identifiers = proof.identifiers[proofIndex]
  if (!identifiers) {
    throw new Error('Invalid indy proof data')
  }
  return {
    schemaId: identifiers.schema_id,
    credentialDefinitionId: identifiers.cred_def_id,
    revocationRegistryId: identifiers.rev_reg_id,
    timestamp: identifiers.timestamp,
  }
}

/*
 * Process indy proof and return
 *  - shared attributes
 *  - shared attribute groups
 *  - resolved predicates
 *  - missing attributes
 *  - missing attribute groups
 *  - unresolved predicates
 * */
export const parseIndyProof = (request: IndyProofRequest, proof: IndyProof): ParsedIndyProof => {
  const result = new ParsedIndyProof()

  for (const [referent, requested_attribute] of Object.entries(request.requested_attributes)) {
    if (requested_attribute.name) {
      const shared = proof.requested_proof.revealed_attrs[referent]
      if (shared) {
        const identifiers = getProofIdentifiers(proof, shared.sub_proof_index)
        result.sharedAttributes.push({
          name: requested_attribute.name,
          value: shared.raw,
          identifiers,
        })
      } else {
        result.unresolvedAttributes.push({
          name: requested_attribute.name,
        })
      }
    }

    if (requested_attribute.names) {
      const shared = proof.requested_proof.revealed_attr_groups[referent]
      if (shared) {
        const attributes = Object.entries(shared.values).map(([name, value]) => ({
          name,
          value: value.raw,
        }))
        const identifiers = getProofIdentifiers(proof, shared.sub_proof_index)
        result.sharedAttributeGroups.push({
          attributes,
          identifiers,
        })
      } else {
        result.unresolvedAttributeGroups.push(requested_attribute.names.map((name) => ({ name })))
      }
    }
  }

  for (const [referent, requestedPredicate] of Object.entries(request.requested_predicates)) {
    // @ts-ignore Mistake in AFJ type definition
    const shared = proof.requested_proof.predicates[referent]
    if (shared) {
      const identifiers = getProofIdentifiers(proof, shared.sub_proof_index)
      result.resolvedPredicates.push({
        name: requestedPredicate.name,
        predicateType: requestedPredicate.p_type,
        predicateValue: requestedPredicate.p_value,
        identifiers,
      })
    } else {
      result.unresolvedPredicates.push({
        name: requestedPredicate.name,
        predicateType: requestedPredicate.p_type,
        predicateValue: requestedPredicate.p_value,
      })
    }
  }

  return result
}

export class CredentialSharedProofData {
  public sharedAttributes: Array<SharedAttribute>
  public sharedAttributeGroups: Array<SharedAttributesGroup>
  public resolvedPredicates: Array<ResolvedPredicate>

  public constructor() {
    this.sharedAttributes = []
    this.sharedAttributeGroups = []
    this.resolvedPredicates = []
  }
}

export type GroupedSharedProofDataItem = { data: CredentialSharedProofData; identifiers: ProofIdentifier }
export type GroupedSharedProofData = Map<string, GroupedSharedProofDataItem>

/*
 * Group parsed indy proof data ( returned from `parseIndyProof`) by credential definition id
 * */
export const groupSharedProofDataByCredential = (data: ParsedIndyProof): GroupedSharedProofData => {
  const result: GroupedSharedProofData = new Map<string, GroupedSharedProofDataItem>()
  for (const item of data.sharedAttributes) {
    if (!result.has(item.identifiers.credentialDefinitionId)) {
      result.set(item.identifiers.credentialDefinitionId, {
        data: new CredentialSharedProofData(),
        identifiers: item.identifiers,
      })
    }
    result.get(item.identifiers.credentialDefinitionId)?.data.sharedAttributes.push(item)
  }
  for (const item of data.sharedAttributeGroups) {
    if (!result.has(item.identifiers.credentialDefinitionId)) {
      result.set(item.identifiers.credentialDefinitionId, {
        data: new CredentialSharedProofData(),
        identifiers: item.identifiers,
      })
    }
    result.get(item.identifiers.credentialDefinitionId)?.data.sharedAttributeGroups.push(item)
  }
  for (const item of data.resolvedPredicates) {
    if (!result.has(item.identifiers.credentialDefinitionId)) {
      result.set(item.identifiers.credentialDefinitionId, {
        data: new CredentialSharedProofData(),
        identifiers: item.identifiers,
      })
    }
    result.get(item.identifiers.credentialDefinitionId)?.data.resolvedPredicates.push(item)
  }
  return result
}

/*
 * Retrieve proof details from AFJ record
 * */
export const getProofData = async (agent: Agent, recordId: string): Promise<ParsedIndyProof | undefined> => {
  const data = await agent.proofs.getFormatData(recordId)
  if (data.request?.indy && data.presentation?.indy) {
    return parseIndyProof(data.request.indy, data.presentation.indy)
  }
  return undefined
}

/*
 * Check if a presentation received
 * */
export const isPresentationReceived = (record: ProofExchangeRecord) => {
  return record.state === ProofState.PresentationReceived || record.state === ProofState.Done
}
