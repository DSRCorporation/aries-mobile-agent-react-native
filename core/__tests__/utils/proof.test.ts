import { parseIndyProof } from '../../App/utils/proof'

const proof_request = {
  name: 'proof-request',
  version: '1.0',
  nonce: '1298236324864',
  requested_attributes: {
    attribute_1: {
      names: ['first_name', 'second_name'],
    },
    attribute_2: {
      name: 'date',
    },
  },
  requested_predicates: {
    predicate_1: {
      name: 'age',
      p_type: '>=' as any,
      p_value: 1,
    },
  },
}

const proof = {
  proof: {
    proofs: [],
    aggregated_proof: {
      c_list: [],
    },
  },
  requested_proof: {
    revealed_attrs: {
      attribute_2: {
        sub_proof_index: 0,
        raw: '1/1/1',
        encoded: '',
      },
    },
    revealed_attr_groups: {
      attribute_1: {
        sub_proof_index: 0,
        values: {
          first_name: {
            raw: 'Aries',
            encoded: '',
          },
          second_name: {
            raw: 'Bifold',
            encoded: '',
          },
        },
      },
    },
    self_attested_attrs: {},
    unrevealed_attrs: {},
    requested_predicates: {
      predicate_1: {
        sub_proof_index: 0,
      },
    },
  },
  identifiers: [],
}

describe('Helpers', () => {
  test('Build indy proof request from template containing two requested attributes', async () => {
    const proofRequest = parseIndyProof(proof_request, proof)
    expect(proofRequest).toMatchSnapshot()
  })
})
