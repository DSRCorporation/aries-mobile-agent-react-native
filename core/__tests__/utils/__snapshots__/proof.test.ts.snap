// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers Build indy proof request from template containing two requested attributes 1`] = `
Object {
  "resolvedPredicates": Array [
    Object {
      "name": "age",
      "predicateType": ">=",
      "predicateValue": 1,
    },
  ],
  "sharedAttribute": Array [
    Object {
      "name": "date",
      "value": "1/1/1",
    },
  ],
  "sharedAttributeGroups": Array [
    Array [
      Object {
        "name": "first_name",
        "value": "Aries",
      },
      Object {
        "name": "second_name",
        "value": "Bifold",
      },
    ],
  ],
  "unresolvedAttributeGroups": Array [],
  "unresolvedAttributes": Array [],
  "unresolvedPredicates": Array [],
}
`;
